bionic/code_references.py

```diff
@@ -0,0 +1,191 @@
+    the returned reference would be ``"call"`` instead of ``MyClass.call``.
+
+    .. code-block:: python
+
+        def x():
+            my_cls = get_my_class("MyClass") # Returns class MyClass
+            my_cls.call()
+
+
+    On a higher level, for Smart Caching, this leaves some changes that Bionic
+    won't be able to detect. Bionic should be able to detect any changes except
+    when a function returns a class or a module. In those case, Bionic cannot
+    detect what is returned. But it should be able to detect the class / module
+    returned when hashing the function which returns the said result. Bionic will
+    still hash those classes / modules, and detect as many changes as it can in
+    them.
+    """
+    # Top of the stack.
+    tos = None
+    lineno = None
+    refs = []
+
+    def set_tos(t):
+        nonlocal tos
+        if tos is not None:
+            # If the top of stack item already exists, that means we
+            # have gone through all the instructions that use the item,
+            # and it is a reference object.
+            refs.append(tos)
+        tos = t
+
+    # Our goal is to find referenced objects. The problem is that co_names
+    # does not have full qualified names in it. So if you access `foo.bar`,
+    # co_names has `foo` and `bar` in it but it doesn't tell us that the
+    # code reads `bar` of `foo`. We are going over the bytecode to resolve
+    # from which object an attribute is requested.
+    # Read more about bytecode at https://docs.python.org/3/library/dis.html
+
+    for op in dis.get_instructions(code):
+        try:
+            # Sometimes starts_line is None, in which case let's just remember the
+            # previous start_line (if any). This way when there's an exception we at
+            # least can point users somewhat near the line where the error stems from.
+            if op.starts_line is not None:
+                lineno = op.starts_line
+
+            if op.opname in ["LOAD_GLOBAL", "LOAD_NAME"]:
+                if op.argval in context.globals:
+                    set_tos(context.globals[op.argval])
+                else:
+                    set_tos(op.argval)
+            elif op.opname in ["LOAD_DEREF", "LOAD_CLOSURE"]:
+                set_tos(context.cells[op.argval])
+            elif op.opname == "IMPORT_NAME":
+                # Streamlit catches ImportError and returns op.argval
+                # as the reference instead. If we see errors with
+                # importing modules, consider adding that check.
+                set_tos(importlib.import_module(op.argval))
+            elif op.opname in ["LOAD_METHOD", "LOAD_ATTR", "IMPORT_FROM"]:
+                if tos is None:
+                    refs.append(op.argval)
+                elif isinstance(tos, str):
+                    tos += "." + op.argval
+                else:
+                    tos = getattr(tos, op.argval)
```

Comment:
I don't think it's possible for attribute to not exist. Python should throw an error during compile time in that case.

But yes, if the attribute is a property, it will run the property method code. I didn't find a way to detect whether a variable is a property.

Reply:
I was very off mark about python compile time checks. For now, I'm going to gate this behind `hasattr` check.

We also discussed this off-GitHub but we might end up calling `getattr` only for modules. But that would require some more thought to make sure we don't miss anything. I'll leave that as a TODO.

Reply:
I'm gonna leave a TODO for properties too.